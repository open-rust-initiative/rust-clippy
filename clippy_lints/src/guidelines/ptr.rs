use std::ops::ControlFlow;

use clippy_utils::consts::{constant, Constant};
use clippy_utils::visitors::for_each_expr;
use clippy_utils::{
    fn_def_id, is_integer_literal, is_lint_allowed, is_path_diagnostic_item, path_res, peel_hir_expr_while,
};
use rustc_hir::def::Res;
use rustc_hir::def_id::DefIdSet;
use rustc_hir::hir_id::HirId;
use rustc_hir::{Block, BorrowKind, Expr, ExprKind, Local, Node, TyKind, UnOp};
use rustc_lint::LateContext;
use rustc_span::symbol::sym;
use rustc_span::Span;

use if_chain::if_chain;

use super::DANGLING_PTR_DEREFERENCE;
use super::NULL_PTR_DEREFERENCE;
use super::PTR_DOUBLE_FREE;

macro_rules! span_ptr_lint {
    ($cx:expr, $lint:ident, $span:expr, $note_span:expr) => {{
        let msg_and_note_msg = match stringify!($lint) {
            "NULL_PTR_DEREFERENCE" => Some(("dereferencing null pointer", "first dereference occurred here")),
            "DANGLING_PTR_DEREFERENCE" => Some((
                "dereferencing a raw pointer that was already freed",
                "the pointer was freed here",
            )),
            "PTR_DOUBLE_FREE" => Some(("pointer was freed multiple times", "second free occurred here")),
            _ => None,
        };
        if let Some((msg, note_msg)) = msg_and_note_msg {
            clippy_utils::diagnostics::span_lint_and_note($cx, $lint, $span, msg, Some($note_span), note_msg);
        }
    }};
}

pub(super) fn check_assign<'tcx, 'a: 'tcx>(cx: &LateContext<'tcx>, expr: &'a Expr<'tcx>) {
    if_chain! {
        if !is_lint_allowed(cx, NULL_PTR_DEREFERENCE, expr.hir_id);
        if !expr.span.from_expansion();
        if let Some(mut ptr_validator) = PtrValidator::from_assignment(cx, expr);
        then {
            ptr_validator.visit(cx);
            if let Some(deref_span) = ptr_validator.deref_span {
                span_ptr_lint!(cx, NULL_PTR_DEREFERENCE, expr.span, deref_span);
            }
        }
    }
}

pub(super) fn check_local<'tcx, 'a: 'tcx>(cx: &LateContext<'tcx>, local: &'a Local<'tcx>) {
    if_chain! {
        if !is_lint_allowed(cx, NULL_PTR_DEREFERENCE, local.hir_id);
        if !local.span.from_expansion();
        if let Some(mut ptr_validator) = PtrValidator::from_init(cx, local);
        then {
            ptr_validator.visit(cx);
            if let Some(deref_span) = ptr_validator.deref_span {
                span_ptr_lint!(cx, NULL_PTR_DEREFERENCE, local.span, deref_span);
            }
        }
    }
}

pub(super) fn check_call(cx: &LateContext<'_>, call_expr: &Expr<'_>, free_fns: &DefIdSet) {
    if !is_lint_allowed(cx, DANGLING_PTR_DEREFERENCE, call_expr.hir_id)
        || !is_lint_allowed(cx, PTR_DOUBLE_FREE, call_expr.hir_id)
    {
        let Some(mut ptr_validator) = PtrValidator::from_call(cx, call_expr, free_fns) else { return };

        ptr_validator.visit(cx);

        if let Some(deref_span) = ptr_validator.deref_span {
            span_ptr_lint!(cx, DANGLING_PTR_DEREFERENCE, deref_span, call_expr.span);
        }
        if let Some(free_span) = ptr_validator.free_span {
            span_ptr_lint!(cx, PTR_DOUBLE_FREE, call_expr.span, free_span);
        }
    }
}

struct PtrValidator<'a, 'tcx> {
    /// HirId of this pointer variable
    hir_id: HirId,
    parent_block: &'a Block<'tcx>,
    search_span: Span,
    is_null: bool,
    /// This should be the span where this pointer gets dereferenced.
    deref_span: Option<Span>,
    /// True if this validator was initialized from a `free` function call.
    was_freed: bool,
    /// This should be the span where this pointer gets freed.
    free_span: Option<Span>,
    /// The function id set for deallocation functions,
    /// it's used for identifying where this pointer got freed.
    free_fns: Option<&'a DefIdSet>,
}

impl<'a, 'tcx> PtrValidator<'a, 'tcx> {
    /// From a `.. = ...` expression, `i.e.: x = null()`
    fn from_assignment(cx: &'a LateContext<'tcx>, expr: &'a Expr<'tcx>) -> Option<Self> {
        if let ExprKind::Assign(assign_to, assigning, _) = expr.kind &&
            let Res::Local(hir_id) = path_res(cx, assign_to)
        {
            let parent_block = get_parent_block(cx, expr.hir_id)?;
            let assign_null = expr_is_creating_null_ptr(cx, assigning);
            Some(Self::new(expr.span, hir_id, assign_null, parent_block))
        } else {
            None
        }
    }
    /// From a `let .. = ...` pattern, `i.e.: let x: *mut i8 = null()`
    fn from_init(cx: &'a LateContext<'tcx>, local: &'a Local<'tcx>) -> Option<Self> {
        if let Local {
            pat,
            init: Some(init_expr),
            ..
        } = local
        {
            let parent_block = get_parent_block(cx, local.hir_id)?;
            let declared_null = expr_is_creating_null_ptr(cx, init_expr);
            Some(Self::new(local.span, pat.hir_id, declared_null, parent_block))
        } else {
            None
        }
    }
    /// From a call to free functions, where the free functions are defined by user config
    fn from_call(cx: &'a LateContext<'tcx>, call_expr: &Expr<'_>, free_fns: &'a DefIdSet) -> Option<Self> {
        // FIXME: It might not always the case that the first parameter is the pointer got freed,
        // we need a more ergonomic solution.
        if let ExprKind::Call(_, [first_param, ..]) = call_expr.kind &&
            let Res::Local(hir_id) = path_res(cx, first_param)
        {
            let parent_block = get_parent_block(cx, call_expr.hir_id)?;
            let assumed_null = false;
            Some(
                Self::new(call_expr.span, hir_id, assumed_null, parent_block)
                    .was_freed(true)
                    .free_fns(free_fns)
            )
        } else {
            None
        }
    }

    fn new(span: Span, path_hir_id: HirId, is_null: bool, parent_block: &'a Block<'tcx>) -> Self {
        // This will produce:
        // { let ptr = std::ptr::null(); do_something(); }
        //   --------------------------: span
        // -----------------------------------------------: parent_block
        //                              ------------------: search_span
        let search_span = parent_block.span.with_lo(span.hi());
        Self {
            hir_id: path_hir_id,
            parent_block,
            search_span,
            is_null,
            deref_span: None,
            was_freed: false,
            free_span: None,
            free_fns: None,
        }
    }

    fn was_freed(mut self, yes: bool) -> Self {
        self.was_freed = yes;
        self
    }

    fn free_fns(mut self, ids: &'a DefIdSet) -> Self {
        self.free_fns = Some(ids);
        self
    }

    fn visit(&mut self, cx: &'a LateContext<'tcx>) {
        let _res: Option<()> = for_each_expr(self.parent_block, |ex| {
            // Skipping irrelevant span
            if !self.search_span.contains(ex.span) {
                return ControlFlow::Continue(());
            }
            match ex.kind {
                // check for `ptr = non_null()` and `*ptr = some_value()`
                ExprKind::Assign(assign_to, assigning, _) => {
                    if path_hir_id_matches_id(cx, assign_to, self.hir_id) {
                        self.is_null = expr_is_creating_null_ptr(cx, assigning);
                        self.was_freed = false;
                    }
                },
                ExprKind::Unary(UnOp::Deref, expr) => {
                    if path_hir_id_matches_id(cx, expr, self.hir_id) &&
                        (self.is_null || self.was_freed)
                    {
                        self.deref_span = Some(ex.span);
                        return ControlFlow::Break(());
                    }
                },
                // FIXME: It might not always the case that the first parameter is the pointer got freed,
                // we need a more ergonomic solution.
                ExprKind::Call(_ , [first_arg, ..]) => {
                    if path_hir_id_matches_id(cx, first_arg, self.hir_id) &&
                        let Some(free_fn_ids) = self.free_fns &&
                        let Some(fn_did) = fn_def_id(cx, ex) &&
                        free_fn_ids.contains(&fn_did) &&
                        self.was_freed
                    {
                        self.free_span = Some(ex.span);
                        return ControlFlow::Break(());
                    }
                },
                _ => {
                    if path_hir_id_matches_id(cx, ex, self.hir_id) {
                        self.is_null = false;
                    }
                },
            }
            ControlFlow::Continue(())
        });
    }
}

/// Checks the given expression to see if its a:
///
/// 1. `std::ptr::null` or `std::ptr::null_mut` call.
/// 2. Integer `0` casting as a raw pointer.
/// 3. Constant that can be resolved as null pointer.
fn expr_is_creating_null_ptr(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
    match expr.kind {
        ExprKind::Path(_) if matches!(constant(cx, cx.typeck_results(), expr), Some((Constant::RawPtr(0), _))) => true,
        ExprKind::Cast(inner_expr, cast_ty)
            if is_integer_literal(peel_casts(inner_expr), 0) && matches!(cast_ty.kind, TyKind::Ptr(_)) =>
        {
            true
        },
        ExprKind::Call(call, [])
            if is_path_diagnostic_item(cx, call, sym::ptr_null)
                || is_path_diagnostic_item(cx, call, sym::ptr_null_mut) =>
        {
            true
        },
        _ => false,
    }
}

/// Peels all casts and return the inner most (non-cast) expression.
///
/// i.e.
///
/// ```
/// some_expr as *mut i8 as *mut i16 as *mut i32 as *mut i64
/// ```
///
/// Will return expression for `some_expr`.
fn peel_casts<'a, 'tcx>(maybe_cast_expr: &'a Expr<'tcx>) -> &'a Expr<'tcx> {
    if let ExprKind::Cast(expr, _) = maybe_cast_expr.kind {
        peel_casts(expr)
    } else {
        maybe_cast_expr
    }
}

// FIXME: this will have FN if the pointer was declared globally,
// i.e. `const PTR: *mut i8 = 0 as *mut _;`, which should be fine at the moment,
// since its hard to detect dereferecing on a global null pointer without the help of mir.
fn get_parent_block<'tcx, 'a>(cx: &LateContext<'tcx>, id: HirId) -> Option<&'a Block<'tcx>> {
    for (_, node) in cx.tcx.hir().parent_iter(id) {
        if let Node::Block(block) = node {
            return Some(block);
        }
    }
    None
}

fn path_hir_id_matches_id(cx: &LateContext<'_>, maybe_path: &Expr<'_>, id: HirId) -> bool {
    let peeled = peel_hir_expr_while(maybe_path, |e| match e.kind {
        ExprKind::AddrOf(BorrowKind::Ref, _, e) | ExprKind::Unary(_, e) => Some(e),
        _ => None,
    });

    matches!(path_res(cx, peeled), Res::Local(hid) if hid == id)
}
