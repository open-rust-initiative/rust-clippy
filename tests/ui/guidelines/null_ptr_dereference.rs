#![allow(unused)]
#![allow(clippy::zero_ptr)]
#![allow(clippy::needless_late_init)]
#![warn(clippy::null_ptr_dereference)]

use std::ptr::{null, null_mut};

struct Pointer {
    inner: i8,
}

impl Pointer {
    fn as_ptr(&self) -> *const i8 {
        (&self.inner) as *const _
    }
}

fn do_something(_p: i8) {}

fn init_null_ptrs() {
    let a: *mut Pointer = null_mut();
    unsafe {
        do_something((*a).inner);
    }
}

fn assign_null_ptrs() {
    let a: *mut Pointer;
    a = null_mut();
    unsafe {
        do_something((*a).inner);
    }

    let mut b: *mut Pointer;
    b = 0 as *mut Pointer;
    unsafe {
        do_something((*b).inner);
    }

    let mut reassign_later: *const Pointer = 1 as *const _;
    reassign_later = null();
    let _ = unsafe { (*reassign_later).inner };
    reassign_later = 0 as *const _;
    unsafe { if (*reassign_later).inner > 0 {} }
}

fn deref_in_macros() {
    let a: *mut Pointer = 0 as *mut _;
    unsafe {
        println!("{}", (*a).inner);
    }

    let foo_ptr: *mut Pointer = std::ptr::null_mut();
    unsafe {
        println!("{:?}", (*foo_ptr).inner);
    }
}

unsafe fn misc_deref() {
    let mut still_null: *mut i8 = null_mut();
    still_null = 0 as *mut i8;
    let _ = *still_null;

    let as_train: *mut i8;
    as_train = 0 as *mut i32 as *mut u16 as *mut i8;
    let _ = *as_train;
}

// Don't lint anything in this function
// There might be seg-fault in actual code, but it's ok, because we are not testing that.
unsafe fn deref_non_null() {
    let mut as_p: *mut Pointer = 0 as *mut _;
    let mut null_p: *const i8 = null();
    let mut null_mut_p: *mut i8 = null_mut();
    let mut null_pointer: *const Pointer = null();

    fn assign_pointer(mut p: *const Pointer, pointer: &Pointer) {
        p = pointer.as_ptr() as *const Pointer;
    }

    as_p = 1 as *mut _;
    println!("{:?}", (*as_p).inner);

    let good_pointer = Pointer { inner: 1 };
    null_p = good_pointer.as_ptr();
    println!("{:?}", *null_p);

    null_mut_p = 1 as *mut i32 as *mut i16 as *mut i8;
    println!("{:?}", *null_mut_p);

    assign_pointer(null_pointer, &good_pointer);
    println!("{:?}", (*null_pointer).inner);
}

// Don't anything in this function
unsafe fn checked_deref() {
    let as_p: *mut Pointer = 0 as *mut _;
    let null_p: *const i8 = null();

    if !as_p.is_null() {
        println!("{:?}", (*as_p).inner);
    }
    if !null_p.is_null() {
        do_something(*null_p);
    }
}

unsafe fn wrongly_checked_deref() {
    let null_0: *const i8 = null();
    let null_o: *const i8 = null();

    if !null_0.is_null() {
        do_something(*null_o);
    }
    if !null_o.is_null() {
        do_something(*null_0); // FIXME: False negative
    }
}

#[allow(clippy::no_effect)]
unsafe fn override_value() {
    let null_p: *mut i8 = null_mut();
    *null_p = 1_i8;
    *null_p;
}

fn main() {}
