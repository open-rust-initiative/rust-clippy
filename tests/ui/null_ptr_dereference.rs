#![allow(unused)]
#![warn(clippy::null_ptr_dereference)]

fn main() {
    // test code goes here
}
