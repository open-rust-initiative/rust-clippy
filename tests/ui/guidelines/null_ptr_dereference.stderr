error: dereferencing null pointer
  --> $DIR/null_ptr_dereference.rs:21:5
   |
LL |     let a: *mut Pointer = null_mut();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: first dereference occurred here
  --> $DIR/null_ptr_dereference.rs:23:22
   |
LL |         do_something((*a).inner);
   |                      ^^^^
   = note: `-D clippy::null-ptr-dereference` implied by `-D warnings`

error: dereferencing null pointer
  --> $DIR/null_ptr_dereference.rs:29:5
   |
LL |     a = null_mut();
   |     ^^^^^^^^^^^^^^
   |
note: first dereference occurred here
  --> $DIR/null_ptr_dereference.rs:31:22
   |
LL |         do_something((*a).inner);
   |                      ^^^^

error: dereferencing null pointer
  --> $DIR/null_ptr_dereference.rs:35:5
   |
LL |     b = 0 as *mut Pointer;
   |     ^^^^^^^^^^^^^^^^^^^^^
   |
note: first dereference occurred here
  --> $DIR/null_ptr_dereference.rs:37:22
   |
LL |         do_something((*b).inner);
   |                      ^^^^

error: dereferencing null pointer
  --> $DIR/null_ptr_dereference.rs:41:5
   |
LL |     reassign_later = null();
   |     ^^^^^^^^^^^^^^^^^^^^^^^
   |
note: first dereference occurred here
  --> $DIR/null_ptr_dereference.rs:42:22
   |
LL |     let _ = unsafe { (*reassign_later).inner };
   |                      ^^^^^^^^^^^^^^^^^

error: dereferencing null pointer
  --> $DIR/null_ptr_dereference.rs:43:5
   |
LL |     reassign_later = 0 as *const _;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: first dereference occurred here
  --> $DIR/null_ptr_dereference.rs:44:17
   |
LL |     unsafe { if (*reassign_later).inner > 0 {} }
   |                 ^^^^^^^^^^^^^^^^^

error: dereferencing null pointer
  --> $DIR/null_ptr_dereference.rs:48:5
   |
LL |     let a: *mut Pointer = 0 as *mut _;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: first dereference occurred here
  --> $DIR/null_ptr_dereference.rs:50:24
   |
LL |         println!("{}", (*a).inner);
   |                        ^^^^

error: dereferencing null pointer
  --> $DIR/null_ptr_dereference.rs:53:5
   |
LL |     let foo_ptr: *mut Pointer = std::ptr::null_mut();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: first dereference occurred here
  --> $DIR/null_ptr_dereference.rs:55:26
   |
LL |         println!("{:?}", (*foo_ptr).inner);
   |                          ^^^^^^^^^^

error: dereferencing null pointer
  --> $DIR/null_ptr_dereference.rs:61:5
   |
LL |     still_null = 0 as *mut i8;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: first dereference occurred here
  --> $DIR/null_ptr_dereference.rs:62:13
   |
LL |     let _ = *still_null;
   |             ^^^^^^^^^^^

error: dereferencing null pointer
  --> $DIR/null_ptr_dereference.rs:65:5
   |
LL |     as_train = 0 as *mut i32 as *mut u16 as *mut i8;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: first dereference occurred here
  --> $DIR/null_ptr_dereference.rs:66:13
   |
LL |     let _ = *as_train;
   |             ^^^^^^^^^

error: dereferencing null pointer
  --> $DIR/null_ptr_dereference.rs:110:5
   |
LL |     let null_o: *const i8 = null();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: first dereference occurred here
  --> $DIR/null_ptr_dereference.rs:113:22
   |
LL |         do_something(*null_o);
   |                      ^^^^^^^

error: aborting due to 10 previous errors

